EIPS/eip-4762.md

```diff
@@ -35,155 +35,266 @@ def get_storage_slot_tree_keys(storage_key: int) -> [int, int]:
+  index, suffix = get_storage_slot_key(contract_address, storage_key)
+  add_read_access_event(contract_address, index, suffix)
+  charge_gas()
 ```
 
- * At each step of EVM execution, if and only if `PC < len(code)`, chunk `PC // CHUNK_SIZE` (where `PC` is the current program counter) of the callee is accessed. In particular, note the following corner cases:
-     * The destination of a `JUMP` (or positively evaluated `JUMPI`) is considered to be accessed, even if the destination is not a jumpdest or is inside pushdata
-     * The destination of a `JUMPI` is not considered to be accessed if the jump conditional is `false`.
-     * The destination of a jump is not considered to be accessed if the execution gets to the jump opcode but does not have enough gas to pay for the gas cost of executing the `JUMP` opcode (including chunk access cost if the `JUMP` is the first opcode in a not-yet-accessed chunk)
-     * The destination of a jump is not considered to be accessed if it is beyond the code (`destination >= len(code)`)
-     * If code stops execution by walking past the end of the code, `PC = len(code)` is not considered to be accessed
- * If the current step of EVM execution is a `PUSH{n}`, all chunks `(PC // CHUNK_SIZE) <= chunk_index <= ((PC + n) // CHUNK_SIZE)` of the callee are accessed.
- * If a nonzero-read-size `CODECOPY` or `EXTCODECOPY` read bytes `x...y` inclusive, all chunks `(x // CHUNK_SIZE) <= chunk_index <= (min(y, code_size - 1) // CHUNK_SIZE)` of the accessed contract are accessed.
-     * Example 1: for a `CODECOPY` with start position 100, read size 50, `code_size = 200`, `x = 100` and `y = 149`
-     * Example 2: for a `CODECOPY` with start position 600, read size 0, no chunks are accessed
-     * Example 3: for a `CODECOPY` with start position 1500, read size 2000, `code_size = 3100`, `x = 1500` and `y = 3099`
- * `CODESIZE`, `EXTCODESIZE` and `EXTCODEHASH` do NOT access any chunks.
-    When a contract is created, access chunks `0 ... (len(code)+30)//31`
+##### Access events for delegation
 
-### Write Events
+When `contract_address` is that of a delegating EoA, process the following events:
 
-We define **write events** as follows. Note that when a write takes place, an access event also takes place (so the definition below should be a subset of the definition of access events). A write event is of the form `(address, sub_key, leaf_key)`, determining what data is being written to.
-
-#### Write events for account headers
-
-When a nonzero-balance-sending `CALL`, `CALLCODE` or `SELFDESTRUCT` with a given sender and recipient takes place, process these write events:
+ - For an `EXTCODEHASH`:
 
 ```
-(caller, 0, BASIC_DATA_LEAF_KEY)
-(callee, 0, BASIC_DATA_LEAF_KEY)
+(delegated_address, 0, CODEHASH_LEAF_KEY)
 ```
 
-if no account exists at `callee_address`, also process:
+ - For an `EXTCODESIZE`, or any other `*CALL`, `CALLCODE`:
 
 ```
-(callee, 0, CODEHASH_LEAF_KEY)
+(delegated_address, 0, BASIC_DATA_LEAF_KEY)
```

Comment:
Should we specify whether or not these event accesses are write or read? The previous EIP mentioned that these are write accesses

Reply:
that part still needs to be pythonized but yes we will